src/components/ModalPortal.js

```diff
@@ -216,6 +217,21 @@ export default class ModalPortal extends Component {
       }
     }
     this.shouldClose = null;
+    this.moveFromContentToOverlay = null;
+  }
+
+  handleOverlayOnMouseUp = () => {
+    if (this.moveFromContentToOverlay === null) {
```

Comment:
Perhaps we can simplify this by inspecting the `event.target` or `event.currentTarget` is inside the `this.content`.
`this.shouldClose = this.content.contains(event.target)`.

Reply:
Nice!
Surely, when the mouse move from content then `mouseUp` on `overlay`, the code `this.content.contains(event.target)` always return `false`. 
Basically, we will need to take more steps to get the false value, however, we can get a better understanding.

Reply:
Ok. You can use `shouldClose` instead of a new property.

Reply:
We can make a nice refactoring about all the mouse events handle. Currently, this flag `shouldClose` has 3 states `null | false | true`. We can improve it by moving the logic from `handleOverlayClick` to `handleOverlayMouseUp`. This way only then this event is fired we actually try to close the modal.

`handleOverlayMouseDown` is the tricky part. With this event, we actually set the `shouldClose` to `true` when the `event.target == event.currentTarget`, this way we know that the click was in the `overlay`.

All the other handles must set `shouldClose` to `false`.

Reply:
Here is the idea.

```js
  handleOverlayOnMouseDown = event => {
    this.shouldClose = event.target == event.currentTarget;
  }

  handleOverlayOnMouseUp = (event) => {
    if (this.shouldClose && this.props.shouldCloseOnOverlayClick) {
      this.shouldClose = false;

      if (this.ownerHandlesClose()) {
        this.requestClose(event);
      } else {
        this.focusContent();
      }
    }
  }
```

Reply:
It's very nice to have the tricky way `handleOverlayOnMouseDown` by checking the `event.target` and `event.currentTarget`. Unfortunately, It does not work.
The reason is `handleOverlayOnMouseDown` will not be called when `mouseDown` on the `Slider`, that `Slider` might `stopPropagation` so the `overlay` receives no events.
Here is how I solve the problem:
`handleOverlayOnMouseDown` does not call when `mouseDown` on `Slider`, I init a new prop here which is `moveFromContentToOverlay` with a `null` value, then in `onMouseUp`, if it is still `null` => `this.shouldClose = false`. `null` value here indicates that users `mouseDown` at `slider` then all other cases should set `moveFromContentToOverlay` to `false`. The `moveFromContentToOverlay` is actually another flag, because the `shouldClose` and its 3 flags `(null|false|true)` have been used for your purposes so I need to use another flag. Perhaps there's a better way to solve this problem instead of initialize a new prop, but I find it no way.

Reply:
@mrsilver256 That's fine. 👍 I'm just trying to avoid handling too much state as possible...but if we can't find a better way, I'll be happy to merge this PR.

Check out this branch, [diasbruno/react-modal:fix/handle-mouse-events](https://github.com/diasbruno/react-modal/tree/fix/handle-mouse-events).

Reply:
👍 Nice!

Reply:
@diasbruno Do you know when will this be merged and released?

Reply:
Before merging this PR, I'm planing to make `v3` official.

Reply:
That would take a long time. I'm actually using React Modal inside my project and also get this error, I'd manually handle this, btw, until this get merged.